File: Excel/900-999/944/944 Challenge.py
import pandas as pd
import networkx as nx

path = "900-999/944/944 ES EF Calculation.xlsx"
input = pd.read_excel(path, usecols="A:C", nrows=21, skiprows=1)
test = pd.read_excel(path, usecols="E:G", nrows=21, skiprows=1).rename(columns=lambda c: c.rstrip(".1"))

df = input.copy()
edges = []
for t, deps in zip(df.Task, df.Dependencies):
    if pd.isna(deps):
        continue
    for d in str(deps).split(","):
        d = d.strip()
        if d:
            edges.append((d, t))

G = nx.DiGraph(edges)
G.add_nodes_from(df.Task)

dur = dict(zip(df.Task, df.Duration))
ES, EF = {}, {}

for t in nx.topological_sort(G):
    ES[t] = max((EF[p] for p in G.predecessors(t)), default=0)
    EF[t] = ES[t] + dur[t]

result = pd.DataFrame({"Task": list(ES), "ES": ES.values(), "EF": EF.values()}).sort_values("Task").reset_index(drop=True)

print(result.equals(test))
# True

File: Excel/900-999/944/944 Challenge.R
library(tidyverse)
library(readxl)
library(igraph)

path <- "900-999/944/944 ES EF Calculation.xlsx"
input <- read_excel(path, range = "A2:C22")
test <- read_excel(path, range = "E2:G22")

edges <- input %>%
  filter(Dependencies != "") %>%
  separate_rows(Dependencies, sep = ",\\s*") %>%
  transmute(from = Dependencies, to = Task)

g <- graph_from_data_frame(edges, vertices = input$Task, directed = TRUE)
ord <- topo_sort(g, mode = "out") %>% as_ids()
dur <- deframe(input %>% select(Task, Duration))

ES <- EF <- set_names(numeric(length(ord)), ord)

for (x in ord) {
  p <- neighbors(g, x, mode = "in") %>% as_ids()
  ES[x] <- if (length(p) == 0) 0 else max(EF[p])
  EF[x] <- ES[x] + dur[x]
}

result = tibble(Task = ord, ES = ES[ord], EF = EF[ord]) %>%
  arrange(Task)

(result == test) %>% all()
# [1] TRUE

igraph::plot.igraph(g, vertex.label = V(g)$name)
